### ISSUE TITLE
cmd/go: exclude vendor dir from matching `...`

### ISSUE THREAD

**Issue by natefinch** [golang/go#19090]
There are some closed issues about this, but I'd like to bring it up again.

Any time you're building an application and you make an update to any package in the repo (of which there are often many), you run `go test ./...` to make sure you haven't broken your program.  This worked fine before vendoring became a thing, but with vendoring, it runs all the tests of vendored code as well.  While you may want to do that once in a while, such as when you update the code in the vendor directory, it's generally not something you want to do every time your code changes (because it shouldn't affect code you depend on at all).

I'd like to suggest that `...` ignore the vendor directory by default.  I almost never want to have any command touching the vendor directory directly - not go vet, not go fmt, not go build, not go test, not go install, not go fix.

If the community feels that there is utility in being able to match the vendor directory, I suggest adding a flag called -vendor (or whatever) that indicates `...` should match the vendor directory.  But by default, I think almost everyone would prefer it not match the vendor directory.

As we start down the path of having an official package management solution, I think vendoring will get more and more common.... and this will become more and more of a pain.

(and yes, I know you can do the `go test $(go list | grep -v vendor)` thing, but that's shell and OS-dependent, and ugly, and just bad UX)

**Comment by mvdan** [comment]
I feel similarly about this issue.

`vendor` could be ignred when walking its parent, so `./...` would ignore it but `./vendor/...` wouldn't.

**Comment by dlsniper** [comment]
I would like to add the votes of 21 people that do not want to have their editor run `go test` against `vendor/` at all https://youtrack.jetbrains.com/issue/GO-2366 and 19 others that don't want `gofmt` to run against `vendor/` as well https://youtrack.jetbrains.com/issue/GO-2412. One could argue that the editor should be smarter and send a list of packages to the tool but then why should users be forced to either have to use `  $(go list | grep -v vendor) `, which mind you is not possible on Windows (WSL does not count as it introduces yet again more work for the user, thus bad UX), or use an editor to achieve this?

Also, please notice tools like `glide`, one popular dependency management tool, where there's even a dedicated command to list packages and omit the vendor directory in order to make life easier.

This is clearly an UX problem that the users should not have to face.

**Comment by ianlancetaylor** [comment]
Previous discussion:

- #14417 cmd/go: go test ./... fails on go 1.6 because dependencies are tested
- #17497 cmd/go: go generate ./... will recurse into vendor
- #17879 Exclude vendor/ from linters

Marking as Go 1.9 for further consideration and decision.

**Comment by dlsniper** [comment]
I would also like to add that this is what a Windows user currently has to write when using Windows CMD to work around the vendor issue ` for /f "" %G in ('go list ./... ^| find /i /v "/vendor/"') do @go test %G ` which is a lot worse than ` go test $(go list | grep -v vendor) `.

**Comment by ernesto-jimenez** [comment]
@natefinch thanks for raising this again.

Regarding the usual argument that "`./...` means all sub packages should avoid making `vendor` a especial case":

1. `vendor` **is** an especial case.
2. `./...` already ignores folders with names starting with `.` or `_`, so there's a precedent supporting this proposal.

**Comment by PaulReiber** [comment]
Are there good reasons why the vendor directory cannot simply be renamed
_vendor?

Kind regards,


*Paul Reiber*http://about.me/reiber

**Comment by dlsniper** [comment]
@PaulReiber see this comment: https://github.com/golang/go/issues/11659#issuecomment-173376808

> "vendor" is not being renamed. That ship has long sailed.

@ianlancetaylor can you please add https://github.com/golang/go/issues/1165 to the list of related issues as hopefully your comment will be more visible than mine. Thank you

**Comment by ianlancetaylor** [comment]
@dlsniper Done, thanks.

**Comment by rakyll** [comment]
This was overwhelmingly mentioned in the tools survey I conducted last year. Combining all feedback I got (individuals, mailing list questions, StackOverflow, etc), this particular problem was one of the top problems that has been mentioned.

**Comment by johnt337** [comment]
A core function like test that would understand vendoring as a dependency and not the target of testing scope would be huge. It's inefficient to write custom regex patterns to include some portions of code and not others. Increasingly so when you want to operate on subsets of the code base. The ./... Operator forces you to glob all go files with the expressed understanding that you will filter out a majority of the results returned wastes cycles during iterative testing. Given this is a solved problem with tools like glide it would seem like a quick win scenario with tangible returns to the go dev community.

**Comment by minux** [comment]
This has be discussed at length and rejected.

I think the real solution is #11193.
What people really don't want is to run unneeded tests.

With proper test caching, it doesn't matter go test ./... tests vendored
packages or not.

**Comment by natefinch** [comment]
That doesn't help for all the rest of the tooling. I don't want go vet complaining about problems in vendored code. I definitely do not want go fmt or go fix *changing* things in vendored code. I don't want go install installing binaries that happen to be in vendored repos. 

The root of it is this: without vendoring, ... never matches dependencies that arent't in your repo, because they're not subdirectories. I just want to maintain that functionality while vendoring.

**Comment by minux** [comment]
Then don't invoke go fmt or go fix on ./..., which is very problematic even
if there aren't any vendored packages.

The core problem is people tend to overuse the ./... pattern. The vendor
problem is just a red herring.

**Comment by dlsniper** [comment]
I've also just got bitten by an issue that ` go vet ` complains about code in the `vendor/` directory but until the upstream changes that code I can't do anything there thus I'm forced to use the workaround instead of not having to deal with it at all.

And I don't think that ` go fmt ./... ` is overused when you want to format the project and make sure it's all consistent inside a pipeline for example.

**Comment by minux** [comment]
My proposal (again) is to introduce negative patterns explicitly.

e.g.
go test ./... -./vendor/...  # using - is preferred, but -exclude also
works.

to exclude any packages matching ./vendor/... from the result of ./....

Changing the behavior of ./... today to ignore vendor packages will only
introduce confusion.
(And there are reasonable uses of ./... to match vendor packages.)

**Comment by sparrc** [comment]
> Then don't invoke go fmt or go fix on ./..., which is very problematic even
if there aren't any vendored packages.

why is this "problematic"? We've never had any problems with it, and we run it on every commit merged into InfluxDB, Telegraf, & Kapacitor.

and I personally don't care if my _dependencies_ are go fmt'ed properly, I only want to run it on my codebase.

> Changing the behavior of ./... today to ignore vendor packages will only
introduce confusion.

My experience has been most people being confused by it not ignoring `vendor/`.

**Comment by mibk** [comment]
> e.g.
go test ./... -./vendor/...  # using - is preferred, but -exclude also
works.

While this concept could definitely be applied to more use cases, excluding `vendor` is the real pain IMO. For other use cases `go test $(go list | grep -v ...)` always does the job.

> Changing the behavior of ./... today to ignore vendor packages will only
introduce confusion.
(And there are reasonable uses of ./... to match vendor packages.)

If we don't want to break backward compatibility and therefore prevent potential confusion, could it be tolerable by both sides to introduce a single-character flag for excluding vendor? It would be only 3 keystrokes more to type without any BC break.

**Comment by hodgesds** [comment]
What about a `.gorc` file that can be configured with defaults?

**Comment by minux** [comment]
configuration files are out of the question. the go command is designed to
avoid configuration files in the first place.

Also, whether to ignore a package is property of a specific invocation of
the go command.

Please note that we do want to build/install the vendored packages, so only
ignoring them in test, vet, fmt, but not build looks pretty inconsistent to
me.
That's why I propose adding negative patterns to make it explicit.


To put it another way:

should the ./... pattern matching different set of packages in the
following commands?
go install ./...
go test ./...

**Comment by mvdan** [comment]
@minux why would you want to explicitly install vendored packages? If they are vendored, they are dependencies and will already be built if `vendor` is ignored.

**Comment by sparrc** [comment]
>Please note that we do want to build/install the vendored packages

@minux, that's not true, in telegraf I only want to build the packages in `vendor/` that telegraf depends on, not every single package in `vendor/`.

Building every package in `vendor/` means that you would have to trim out packages/tests that your project doesn't use.

This is not how it works when the project is in your GOPATH.

**Comment by minux** [comment]
Whether to include unnecessary packages under vendor depends on how you
vendor a package (e.g. copy or git submodule), and that's not relevant in
this discussion.

If you vendor only necessary packages, then you could use go install ./....

**Comment by mvdan** [comment]
> If you vendor only necessary packages, then you could use go install ./....

True, but it still doesn't explain why you need the vendored packages to be included. This just explains a way for it to not matter.

**Comment by davecheney** [comment]
Rather than expressing incredulity that people want to use the glob operator, I think it is important to recognise that globbing is a straight forward solution to _very_ commonplace questions like _"How can i run the tests for all the packages in my project?", which, until go 1.6, was satisfied by globbing. 

Saying that people shouldn't use the glob operator is not helpful as the replacement was non portable (how do you pipe grep on windows?), or relies on an external tool like go test $(glide nv); which also isn't a portable solution. 

As @rakyll showed from her research this issue is a major point of confusion for Go developers and that alone should be cause to reexamine the decision.

**Comment by iand** [comment]
Another consideration is that the current behaviour requires you to vendor dependencies of your dependency's tests, e.g. packages like testify

**Comment by jsternberg** [comment]
Related to the above consideration, `go get ./...` will not bring in those dependencies automatically so you have to `go get ./...`, vendor the dependencies, then repeat until no new vendored dependencies show up. While it's easy enough to make a custom tool that does this, it seems like it should be unnecessary for the default go toolchain to cause this problem.

It seems to me like you should be able to call `go get ./...`, vendor the dependencies identifies and retrieved by the command, and `go test ./...` would still work. This isn't currently true.

**Comment by davecheney** [comment]
@jsternberg if possible I'd like to leave `go get ./...` out of this debate. I want everyone to remain laser focused on the usability issue of `go test ./...` matching vendored code.

There are certainly other issues with the Go 1.6+ behaviour, but for the sake of clarity, can those debates be taken elsewhere.

Thanks

**Comment by natefinch** [comment]
@davecheney I actually don't want to limit this just to go test.  While I agree that go get is a special flower, almost none of the normal go tools would seem to be appropriate to run on the vendor directory.

**Comment by slrz** [comment]
@natefinch Why not? The list command comes to mind immediately.

@davecheney How can they be separated? Are you suggesting the possibility of making the glob work differently between test and get? Ugh. Is there any precedent for this in the existing code?

**Comment by davecheney** [comment]
I just want people to focus on the problem of go test _in this issue_.

There are other use cases, they should be argued separately as in my experience if you ask for several things at once, you risk getting none of them.

**Comment by StabbyCutyou** [comment]
I would agree that focusing this only on the UX issues faced by `go test` is the best move for this proposal. 

That said, longer term it would be great if the behavior change becomes standard in other places where the ./... syntax is supported.

But i'd like to see us fix the leaky pipe before trying to refinish the basement. I run into this all the time in many different projects and the lack of a simple solution built into the test tool itself has long been a pain point for myself and many others I know.

**Comment by nathany** [comment]
This is a concern across several tools, not just `go test ./...`. A holistic solution would be appreciated, even if this serves as a meta-issue with separate issues specific to each tool.

As mentioned previously in @rakyll's survey, my main issue is that there isn't a consistent workaround for ignoring the ./vendor folder with the variety of Go tools.

The gofmt tool requires a list of packages, it doesn't work with a `$(go list ./... | grep -v vendor)` exclusion.

The result is a Makefile with a bunch of different commands instead of something simple.

```
GO_PACKAGES = $(shell go list ./... | grep -v vendor)
GO_FILES = $(shell find . -name "*.go" | grep -v vendor | uniq)

gofmt -s -l -w $(GO_FILES)
go vet $(GO_PACKAGES)
go list ./... | grep -v /vendor/ | xargs -L1 golint
go test $(GO_PACKAGES)
```

It took a good amount of time to find the quirks and come up with commands (we're not all bash experts). Multiply that by every team out there running into this. As pointed out in this thread, it gets even worse, in that these commands are not cross-platform.

Specifying `./...` and perhaps a `-novendor` flag across any tool would be simple.

If there is a flag, what it's named, and whether the default is to include vendor or not, I don't have an opinion right now.

**Comment by robbiev** [comment]
I'm not sure if this has been mentioned already: what about supporting a `_vendor` directory _in addition_ to `vendor`?

**Comment by StabbyCutyou** [comment]
If we could get the issue addressed in `test` first, get the behavior correct and make sure it's working as intended, then i'd imagine it would be easier and more palatable to propose rolling the changes to the `vet` tool and others.

If we get it fixed just in `test` or globally, however, i'll be happy either way.

**Comment by cznic** [comment]
First, we've got another special treatment of (yet another) special name the go tool has to handle differently.

Secondly, we saw it has unforeseen consequences.

Thirdly, we propose to fix that by adding another specially treated case.

#Featuritis

**Comment by davecheney** [comment]
> Thirdly, we propose to fix that by adding another specially treated case.

I argued at the time, and will do so again that the vendor directory should have been called `_vendor/`. This would have had less impact on tooling

**Comment by dlsniper** [comment]
I don't understand why we can't consider this as a quality of life improvement based on concrete facts and evidence. It's clear that people don't want tools that use ` ./... ` to touch ` vendor/` at all most of the time.

The question should be: Does it have to be implicit or explicit to omit it? Not if we should do it or it. There's clear evidence that some users have a lot harder time around this missing functionality just by looking at the Windows users and the horrible workaround they have to do.

More evidence this is needed? Hang out long enough in Gophers Slack, read the various mailing lists / issue trackers of various tools or do a [Google Search for `"go test" exclude vendor`](https://www.google.com/search?q=%22go+test%22+exclude+vendor) to find the 10k+ results around this (can't validate them all but they are there if someone wants to tackle this).

My gut feeling is that this option should be explicit in usage so that the current behavior is not changed for the tools we have and be an opt-in feature, as it should have been from the beginning to include the vendor.

**Comment by ScottBrooks** [comment]
I feel like the issue can be better understood as

"After adding vendor support, there is no easy/cross platform way to test just **my** code"

`./...` is not intuitive either, it just happened to be what worked.

Adding the following option to `go test` would avoid needing to change how `./...` works, and would potentially work cross platform.

         -l
             Run the tests on your local files only, excluding the vendor directory.

**Comment by franciscocpg** [comment]
I agree with @disniper and @ScottBrooks. It's better (less confusing) to add a new flag and keep backwards compatibility than changing default behavior.

**Comment by sparrc** [comment]
just want to re-iterate that this should not only apply to `go test`. IMO it would be even more confusing if one of the go tools behaved differently than the others, and I don't think the anybody wants to `fmt` their vendored dependencies either.

**Comment by spenczar** [comment]
I am confused by the arguments against this which object to this on the grounds that it is "yet another special case." `vendor` _is_ a special directory. It's just incompletely implemented as such today. Some tools treat it as special, others do not.

I think this is a question of _consistency_ in understanding that `vendor` is a special directory. That's why people get confused and surprised - they expect the special-ness to be consistently handled.

So: this isn't "yet another" special case. This is the same old one, just completing its implementation.

**Comment by dlsniper** [comment]
@ianlancetaylor / @rsc et.al. involved in the dev team, could you please help / guide us on this one.

Personally, I'd love to see this landing in time for 1.9 if possible, and judging by the number of voters (according to [NoMeToo](https://github.com/golang/go/wiki/NoMeToo) this should be the way to show support for it), I believe that others would be interested in this sooner rather than later.

Thank you for your time and consideration.

**Comment by dlsniper** [comment]
P.S. it doesn't mean the Go dev team has to work on the implementation by all means, just accept / decline it, I'm sure the implementation will follow.

**Comment by smo921** [comment]
Some thoughts from a Go user who finally needed vendoring, and probably has no business even responding to this discussion.  To preface this:  I love the Go community and the language.  I encourage people to try it out.  I've even presented an intro to Go to my team in hopes we would start writing more Go code.  My background is systems operations, not development.  I have however dealt with dependency hell.

The TL;DR: Basically what @spenczar said above.

Now, on to my $0.02.

1. The choice of `vendor` as the directory name means I can no longer have a package named vendor.  It essentially introduced a new reserved word into the language.  I don't foresee needing a package named vendor in my software, but someone might.

2.  The first time using `vendor/` and running `go test ./...` I was met with a slew of errors from my dependencies.  This led to frantic searching for workarounds and advice from the community.  To be 100% honest, I don't care about running vendor tests, they are not code I own, nor are they tests I want to run.  Perhaps I'm naive, I don't write software all day long and generally trust library authors to properly test their code before it is released.

3.  Whatever the fix (there are plenty of good ideas), don't let a desire for 100% backward compatibility  influence the decision.  You've already introduced breaking functionality by using a non-reserved name (`vendor`), instead of something that the golang toolset would ignore by default (`_vendor` or `.vendor` for example)

4. Trying to encourage people to test drive Go without dependency management (ie vendoring) was difficult.  Now that vendoring has been officially adopted the reluctance to make "non-backward compatible" changes to the tool set is frustrating.  It is one more strange edge-case that needs to be explained to newcomers.

My ideal solution would be to have the location of vendor dependencies be ignored by the golang toolset by default.  The subtree should be treated as if it didn't exist or was read-only.  I shouldn't fear that running `go fmt` is going to modify the source code of my dependencies.

Thanks.

**Comment by ianlancetaylor** [comment]
@dlsniper The golang/dep project is overhauling dependency management, I think when that process is completed we will have a better understanding of how to handle `...` and the vendor directory.

**Comment by davecheney** [comment]
Respectfully, dep is an experiment and focused on an entirely different problem; versioning Go packages. 

This issue is about ...  matching vendor/ and the effect is has on developers edit/compile/test inner loop.

I would appreciate it if every problem with the Go tool was not kicked down the road by saying "oh dep'll fix it"

Thanks

Dave

**Comment by dlsniper** [comment]
@ianlancetaylor As per comment here: https://groups.google.com/d/msg/golang-nuts/PaGu2s9knao/Bq4vmFh7AgAJ it seems that golang/dep will not arrive here in 16 months from now as an official tool, and I'm not sure about others, but I'd rather fork the Go frontend and add that option rather than wait that long. 

There are clear signs everyone really do not agree with the current situation and we, the community, want to work with you, the Go team, to get this improved for us, all the Go users.

If creating the CL for this issue is what stops us from having a decision on it, then please indicate so, but given how this is a proposal, even if it doesn't follow the normal proposal flow, I think it must be addressed. Go 1.9 cut-off will be here in almost a month from now, at which point it will be too late for us to change anything for the next few months, which can potentially delay it to Go 1.10, that means, 10 more months to get a `--no-vendor` or `--skip-vendor` or `--omit-vendor` flag.

Thank you for your time and consideration.

**Comment by rsc** [comment]
I wrote in https://github.com/golang/go/issues/11659#issuecomment-171665255:

> I think there are two cases:
>
> 1. A package in a vendor tree that is actually used (directly or
>    indirectly) by the code outside the vendor tree. In this case I think you
>    would want to test the package, just as you test all your other
>    dependencies. If code you depend on is broken, you want to find out in the
>    most precise test possible, not wait until a test of the main program fails
>    mysteriously.
> 2. A package in a vendor tree that is not actually used (directly or
>    indirectly) by the code outside the vendor tree. In this case, obviously no
>    one cares and the test need not run. But in this case I don't understand
>    why tools are bringing that package into the vendor tree in the first
>    place. I would argue that, for many reasons, dead (unused) code should not
>    be copied into a vendor tree. Code that is in your vendor tree has to be
>    maintained, updated, and so on. If it's not needed, it should just be
>    omitted.
> 
> I think it's a mistake to skip the test in (1). And I think (2) indicates a
> problem in the vendoring process.
> 
> The counter-argument to this is that you can't omit individual packages
> when you're using git submodules. But when you're using git submodules I
> think you're tying yourself to the target repository quite a bit more and
> so it may still be appropriate to test the entire thing.

Obviously from the reactions to the issue lots of people are still running into problems. But why? 

Is it because there is dead code in your vendor trees you want to ignore? 

Is it because you don't want to test the vendored code you do use?

**Comment by bradleyfalzon** [comment]
> Is it because you don't want to test the vendored code you do use?

Because you asked: I don't run the tests from the standard library, I don't see why I should run the tests for vendored code either. I want to quickly run my tests, so I can continue developing.

**Comment by cespare** [comment]
At our company we import vendored code without test files, so `go test` is not an issue. But we have to explicitly ignore vendoring whenever we run go vet, golint, or other static analysis tools. We even have  some vendored code that isn't gofmted, sadly, so running a project-wide gofmt also requires ignoring vendor.

**Comment by rsc** [comment]
@cespare, are you running ... wildcard commands for all those, or something else? We're unlikely to change gofmt -r to ignore vendor directories, for example.

**Comment by dlsniper** [comment]
> Is it because you don't want to test the vendored code you do use?

During development or CI it's faster to run only your tests.

If someone decides to change locally a library that they use, then they still want to test the changes, but probably not as part of every build. The test code might be there because of that, or might be there for legal reasons, or might be there because of company policies. 

Test code in vendor/ is not dead code, it's just code that breaks the assumptions on how the tool should behave, slowing down the users by a good margin, making it unfriendly for newcomers and there is no way to override it from within the tool.

**Comment by natefinch** [comment]
I want to easily control when I run tests over vendored code.  The same with gofmt or go vet.  Generally, the code in the vendor directory changes very rarely, so the results of go test or go vet are also going to change very rarely.  Why waste my time running those tests over and over.

And yes, of course we all run wildcard commands.  It's the only reasonable way to run commands over the full repo, and running it over the full repo is the only way to check "have I broken anything with my latest changes?"

There's currently *no way* to run all the tests in a repo and ignore the vendor directory.  If you made the go tool's interpretation of ... ignore the vendor directory by default, then suddenly it just works.  And if you want to run the go tool over the vendor directory, there's only one extra command you have to perform... `cd vendor; go test ./...`    

At worst, this change would make something that is currently impossible possible, and make the current default into a single extra very obvious command that works on all platforms.

**Comment by cespare** [comment]
@rsc yes for the static analysis tools.

`...` isn't really applicable for gofmt, right? (If I run goimports/gofmt manually instead of from my editor, which is rare, I do e.g. `gofmt -l -w .`).

**Comment by rsc** [comment]
@cespare, you brought up gofmt. My point was that changing ... to exclude vendor will only help `go fmt ./...`, not `gofmt .`. (Sorry about the -r, that was a mistake.)

**Comment by SamWhited** [comment]
> A package in a vendor tree that is actually used (directly or
indirectly) by the code outside the vendor tree. In this case I think you
would want to test the package, just as you test all your other
dependencies. If code you depend on is broken, you want to find out in the
most precise test possible, not wait until a test of the main program fails
mysteriously.

Respectfully, I disagree that you would want to run tests on the vendor/ tree for this reason. If we ignore vendor for a moment, I'd like to suggest that generally *only* the public API should be tested. The internal implementation details should be able to be completely rewritten, and as long as the final packages behavior is the same no one should care (and you shouldn't need to ever touch your tests, because behavior is the same). If you aren't touching your public API, and that's the only thing that's tested, you won't have to change your tests even if you completely rewrite your backend.

Similarly, it shouldn't matter what I have vendored or how I'm using it: as long as the behavior of my public API stays the same, I shouldn't have to touch my tests, or worry about tests suddenly failing even though it's an internal detail due to an update that won't affect anything anyways (or, if it does affect something, it should cause *my* tests to fail, not just its own, otherwise my tests are incomplete).

TL;DR — relying on knowing that a vendored package test fails feels like a crutch that is used to not write complete tests for your own package. This should not be the default.

**Comment by davecheney** [comment]
I don't want to run my vendor/d code's tests because often they do not pass
as they require a more elaborate integration test setup, maybe using
programs which are not available on my operating system.

And also, I don't want to run their tests because I trust they did that
before they released their code.

**Comment by natefinch** [comment]
Notably, if Juju moves to use vendoring, including the vendored code with tests would be testing an extra 1.5 million lines of code that you know for sure have not changed since the last time you ran the tests.

**Comment by rsc** [comment]
@smo921, re your (1), importing "x/vendor" is allowed and works just fine, precisely so that you _can_ have a package (or command!) named vendor. What you _can't_ do is import "x/vendor/y".

@SamWhited, the logical conclusion of your argument is that one should only ever write and run end-to-end (integration) tests, never tests of smaller pieces like individual packages. Obviously, I disagree.

@davecheney and @dlsniper, the point @ianlancetaylor was making (after a brief discussion he and I had) is that once the go command has integrated good dependency management, we hope there just won't be that many vendor directories anymore (the same way you never see vendor directories in, say, Rust projects), so this becomes a much smaller or hopefully nonexistent pain point. 

That said, I do appreciate the time scale on that and the significant desire as measured by the emoji reactions to the top post, and I like @bradleyfalzon's analogy. 

I think I have found a reasonable semantics and implementation for keeping ... from matching vendor directories (always; no options), and I'll send it out later this week once I clean up the code and the documentation. (It's not trivial. Most people don't realize how powerful ... is: n...t matches not only net but also net/http/httptest.) Depending on how the CL turns out, maybe we can submit it soon and make a final decision on whether to pull it out during the freeze.

**Comment by davecheney** [comment]
Thanks Russ, this is excellent news.

**Comment by dlsniper** [comment]
@rsc Thank you for this, it's really great news.

**Comment by SamWhited** [comment]
> the logical conclusion of your argument is that one should only ever write and run end-to-end (integration) tests, never tests of smaller pieces like individual packages. Obviously, I disagree.

That comment got away from me a bit, so I may not have been clear, but I certainly wasn't advocating this. It's both not absolute (what ever is?), sometimes you have to test individual little "fiddly bits" (that's a technical term), and it's often possible to write unit tests against your public API too.

**Comment by dlsniper** [comment]
@rsc Thank you for getting the CL done and merged so quickly!

One last small note based on this comment here:

> once the go command has integrated good dependency management, we hope there just won't be that many vendor directories anymore (the same way you never see vendor directories in, say, Rust projects), so this becomes a much smaller or hopefully nonexistent pain point.

Currently the committing ` vendor/ ` directory fills in a very important role in the software life-cycle: it ensures that the dependencies are always there for building / testing the code on the local machine as well as the CI system (helps with the reproducible builds). It prevents cases like ` left-pad ` happening for Go projects.

While I don't know how the future will turn out to be, having the dependencies committed with the application right now is a very important step, especially for companies that want to ensure they are also not dependent on github working or not, or in which their CI doesn't have access to Internet at all.

I can add more but I'd rather not spam others here. Thank you once again.

**Comment by slrz** [comment]
From the CL comments, on the topic of how to get at the previous behaviour following the merge of this.

> You can do something like
>     `go test ./... ./vendor/... ./.../vendor/... ./.../vendor/.../vendor/... `
> depending on how many vendor directories you have in your tree, where they are, whether there are vendors of vendors, and so on

Ugh. I really dislike this change. Excluding vendor from patterns doesn't make any sense to me, as they're defined in terms of directories in the file system. Also, the previous behaviour is pretty useful sometimes (e.g. with `go list`) and modifying it in this way is likely to break at least some users' scripts. Considering that a trivial workaround exists to achieve the same thing, is this change really worth it?

The rationale presented in the CL ("by popular vote", basically) is unconvincing and appears especially weak when compared to the careful thought and design that has gone into other decisions concerning Go's tooling.

**Comment by natefinch** [comment]
Did you read this whole thread? There's a lot of valid reasons here.  And there's no trivial workaround that works on all platforms for the old behavior.

Russ's example of a workaround to get the old behavior is overly complex.  

Most of the time this will work just fine:

`go test ./... ./vendor/...`

It's pretty rare for people to have multiple levels of vendor directories.

Unlike the workaround for the old behavior, this one is cross platform and works everywhere.

**Comment by slrz** [comment]
Yes, I did read all the comments on this issue. I'm not convinced. They mostly boil down to "but it'd be more convenient". That's what shell aliases (or whatever) are for. Adding special cases into the core tooling seems like a way too heavyweight approach for tackling such a problem.

The cross platform argument is a red herring. The workaround doesn't have to function on every system under the sun. It only has to work in the environment it is applied to, be that a developer's shell or that build/test script you're writing.

**Comment by dcheney-atlassian** [comment]
> The cross platform argument is a red herring. The workaround doesn't have to function on every system under the sun. It only has to work in the environment it is applied to, be that a developer's shell or that build/test script you're writing.

But it _does_ have to work on Windows, which are the largest population of Go users, and have neither shell aliases or nice things like grep, or substitution.

**Comment by slrz** [comment]
This hasn't been true for more than a decade. Any supported Windows variant includes a shell that, while somewhat different from Unix shells, is very well capable of such things. I wouldn't be surprised if it even provides a `grep` command by default. It certainly does so for many other common Unix programs (generally as an alias for the native Powershell command with roughly equivalent functionality).

**Comment by sgen** [comment]
So windows users will now have to install a shell they're not familiar with, learn that shell, rewrite their scripts in it, and this is easier than running `go test ./... ./vendor/...`.

**Comment by dlsniper** [comment]
@slrz if you could find something simpler than:

```
for /f "" %G in ('go list ./... ^| find /i /v "/vendor/"') do @go test %G
```

I'd be happy to hear about it.

Meanwhile, this feature was requested by the Go users, which proved with a fairly good technical evidence that not matching ` vendor/ ` in  ` ... ` is desired or at least have a way to disable that behavior.

I believe that the current way that's implemented offers a good balance of that while still allowing ` vendor/ ` to be matched if so needed.

It will rarely be so that you need to type: ` ./... ./vendor/... ./.../vendor/... ` let alone ` ./... ./vendor/... ./.../vendor/... ./.../vendor/.../vendor/... ` (vendor flattening should be a thing by now in any package manager).

Most likely you'd need to type: ` ./... ./vendor/... ` which is still just one character longer than ` --no-vendor ./... ` and one shorter than ` --skip-vendor ./... ` but it's definitely easier to type / remember than the current way to do things.

Also, special cases do exist today, afaik ` testdata ` is one such case (if not only).

So please, try to understand that there are some of us who are really happy to see this, and per the commend here: https://github.com/golang/go/issues/19090#issuecomment-279870835 that number might actually be bigger than you think. So a quality of life improvement for so many people while still allowing for the current behavior to exist is a welcomed change imho.

I hope this helps seeing "the other side" as well.

**Comment by slrz** [comment]
Heck, we're talking about programmers here. They will figure out **something** to exclude items containing a constant substring from a list.

Assuming they even want that particular quirky behaviour instead of the one you'd expect from reading the description of patterns in the help text, without the part about this special case.

**Comment by spenczar** [comment]
@slrz this isn't the quirk. The quirk is that `vendor` is a special directory for the `go build` tool. This _completes_ the implementation of the quirk.

We were in a bad state previously when `vendor` was incompletely implemented. We were in the worst position of all - a quirk that was inconsistent.

If purity doesn't convince you, practicality should: In my experience teaching Go and being a local expert in a large-ish company on Go tools, this was an almost _universal_ stumbling block when people went to lint or test their code. Nobody knew to use `go list` and `grep` ahead of time, and most of them wrinkled their nose and made some disparaging comment about go's tools.

**Comment by cznic** [comment]
@dcheney-atlassian 

> But it does have to work on Windows, which are the largest population of Go users, and have neither shell aliases or nice things like grep, or substitution.

Can you please cite some source of Windows Go users being "*the largest population of Go users*"?

Meanwhile, let me refer to a survey of 3,595 respondents: 

> [When asked which operating systems they develop Go on, 63% of respondents say they use Linux, 44% use MacOS, and 19% use Windows, with multiple choices allowed and 49% of respondents developing on multiple systems. The 51% of responses choosing a single system split into 29% on Linux, 17% on MacOS, 5% on Windows, and 0.2% on other systems.](https://blog.golang.org/survey2016-results#TOC_4.).

**Comment by dcheney-atlassian** [comment]
The windows version is the most popular download of windows/dl. This was reported to me in private correspondence a year or so ago, I don't have access to verify this fact directly.

**Comment by dlsniper** [comment]
Out of respect for everyone, I think we should stop the bike shedding now and maybe open up an issue if we find a problem in the current implementation as asked by @rsc. 

Finally, please note that comments like ` Heck, we're talking about programmers here. ` are really not in the spirit of cooperation / Go. Yes, we are all programmers, but that doesn't mean we all share the same opinion.

The fact that ` vendor/ ` is special is undeniable and if users really really care about traversing that, then maybe they should have to use special scripts instead of those who don't (I've borrowed this last bit from someone who said it in a private discussion).

Lets not have the team lock down the discussion as I consider that to be a bit unfriendly. Thank you.

**Comment by sgen** [comment]
But I want to paint it chartreuse!

**Comment by slrz** [comment]
@dlsniper 

> if you could find something simpler than:
> `for /f "" %G in ('go list ./... ^| find /i /v "/vendor/"') do @go test %G`
> I'd be happy to hear about it.

The following should work in a default Windows system, having installed only the Go MSI package:

`go test $(go list ./... | sls -NotMatch '/vendor/')`

The `sls` ("select string") seems to be the Powershell equivalent to `grep`.

**Comment by AhmetS** [comment]
My soln till this mess gets sorted is with npm package.json. Something like the following:

```
{
  "name": "myapp",
  "version": "1.0.0",
  "scripts": {
    "test:win": "for /f \"\" %G in ('go list ./... ^| find /i /v \"/vendor/\"') do @go test -v %G",
    "test:unix": "go test -v $(go list ./... | grep -v /vendor/)",

    "bench:unix": "go test -bench=. -v $(go list ./... | grep -v /vendor/)",
    "bench:win": "for /f \"\" %G in ('go list ./... ^| find /i /v \"/vendor/\"') do @go test -v -bench=. %G",

    "start:unix": "go run *.go",
    "start:win": "for /f \"\" %G in ('dir /b *.go') do @go run %G"
  }
}
```

**Comment by mantzas** [comment]
better late than never!
